### ISSUE TITLE
Question about UVA sample

### ISSUE THREAD

**Issue by DesmonDay** [dmlc/dgl#3817]
## ❓ Questions and Help

If I want to use uva sample, I must set prefetch feature. Is that right? Otherwise there will be bug like following. (I am using graphsage example)

![截屏2022-03-08 下午6 56 31](https://user-images.githubusercontent.com/20554008/157225034-4974ee2c-49ca-4db2-9a2f-898f5c41cca2.png)

**Comment by jermainewang** [comment]
Are you running the official example? If no, could you please provide the script and necessary information (following the bug report issue format) for us to reproduce?

**Comment by DesmonDay** [comment]
Not the official example.

Code: 

`import torch
import torch.nn as nn
import torch.nn.functional as F
# import torchmetrics.functional as MF
import dgl
import dgl.nn as dglnn
from dgl.data import RedditDataset
import time
import numpy as np
import tqdm
import argparse

class SAGE(nn.Module):
    def __init__(self, in_feats, n_hidden, n_classes):
        super().__init__()
        self.layers = nn.ModuleList()
        self.layers.append(dglnn.SAGEConv(in_feats, n_hidden, 'mean'))
        self.layers.append(dglnn.SAGEConv(n_hidden, n_classes, 'mean'))
        self.dropout = nn.Dropout(0.5)
        self.n_hidden = n_hidden
        self.n_classes = n_classes

    def forward(self, blocks, x):
        h = x
        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
            h = layer(block, h)
            if l != len(self.layers) - 1:
                h = F.relu(h)
                h = self.dropout(h)
        return h


dataset = RedditDataset()
graph = dataset[0]
print(graph.ndata['label'])
train_mask = graph.ndata['train_mask']
val_mask = graph.ndata['val_mask']
test_mask = graph.ndata['test_mask']

train_idx = torch.where(train_mask == True)[0]
valid_idx = torch.where(val_mask == True)[0]

device = 'cuda'
train_idx = train_idx.to(device)
valid_idx = valid_idx.to(device)

model = SAGE(graph.ndata['feat'].shape[1], 256, dataset.num_classes).to(device)
opt = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=5e-4)

# sampler = dgl.dataloading.NeighborSampler([25, 10], prefetch_node_feats=['feat'],
#         prefetch_labels=['label'])
sampler = dgl.dataloading.NeighborSampler([25, 10])
train_dataloader = dgl.dataloading.DataLoader(
        graph, train_idx, sampler, device=device, batch_size=1024, shuffle=True,
        drop_last=False, num_workers=0, use_uva=True)

durations = []
for _ in range(10):
    model.train()
    t0 = time.time()
    for it, (input_nodes, output_nodes, blocks) in enumerate(train_dataloader):
        x = blocks[0].srcdata['feat']
        y = blocks[-1].dstdata['label']
        y_hat = model(blocks, x)
        loss = F.cross_entropy(y_hat, y)
        opt.zero_grad()
        loss.backward()
        opt.step()
        # if it % 20 == 0:
            # acc = MF.accuracy(y_hat, y)
            # mem = torch.cuda.max_memory_allocated() / 1000000
            # print('Loss', loss.item(), 'Acc', acc.item(), 'GPU Mem', mem, 'MB')
    tt = time.time()
    print(tt - t0)
    durations.append(tt - t0)

    """
    model.eval()
    ys = []
    y_hats = []
    for it, (input_nodes, output_nodes, blocks) in enumerate(valid_dataloader):
        with torch.no_grad():
            x = blocks[0].srcdata['feat']
            ys.append(blocks[-1].dstdata['label'])
            y_hats.append(model(blocks, x))
    # acc = MF.accuracy(torch.cat(y_hats), torch.cat(ys))
    print('Validation acc:', acc.item())
    """

print(np.mean(durations[4:]), np.std(durations[4:]))

# Test accuracy and offline inference of all nodes
model.eval()
with torch.no_grad():
    pred = model.inference(graph, device, 4096, 12, graph.device)
    pred = pred[test_idx]
    label = graph.ndata['label'][test_idx]
    # acc = MF.accuracy(pred, label)
    print('Test acc:', acc.item())`


Bug:
<img width="597" alt="截屏2022-03-08 下午8 58 08" src="https://user-images.githubusercontent.com/20554008/157242713-10d4e246-08ee-4c2e-ae83-ff65fda3e288.png">

**Comment by yaox12** [comment]
@DesmonDay This is not a bug. If you don't specify a feature to prefetch, it will remain where it is. For UVA-based sampling, the features are stored in pinned **CPU memory**. So you are required to copy them to the GPU memory manually:
```python
x = blocks[0].srcdata['feat'].to(device)
y = blocks[-1].dstdata['label'].to(device)
```
We recommend setting prefetch feature to let the dataloader handles it automatically for you.

**Comment by DesmonDay** [comment]
I got it, thank you.

**Comment by ayotme** [comment]
> @DesmonDay If you don't specify a feature to prefetch, it will remain where it is. For UVA-based sampling, the features are stored in pinned **CPU memory**. So you are required to copy them to the GPU memory manually:
> 
> ```python
> x = blocks[0].srcdata['feat'].to(device)
> y = blocks[-1].dstdata['label'].to(device)
> ```
> 
> We recommend setting prefetch features to let the dataloader handles them automatically for you.
Is it possible that not setting prefetch features would be faster than setting them when UVA sampling was enabled? In my  case, setting prefetch features explicitly did not decrease the training time but increased instead.

Here is my code:
` # g is one heterograph.
neighbor_sampler = dgl.dataloading.NeighborSampler([16, 4], prefetch_node_feats = node_feat_dict, prefetch_labels = node_feat_dict)
sampler = dgl.dataloading.as_edge_prediction_sampler(sampler=neighbor_sampler)
dataloader = dgl.dataloading.DataLoader(g, {g.etypes[-1]: eids}, sampler, device=torch.device('cuda'),
        batch_size=1024, shuffle=True, drop_last=False, num_workers=0, use_uva=True)
`

**Comment by yaox12** [comment]
@ayotme This could happen, but I think the gap should be small. If you observed a big regression when setting prefetch features, please provide more details.

**Comment by maqy1995** [comment]
Hi @yaox12 , I met the same problem like @ayotme . When I set prefetch features, it will be 25% slower than not to prefetch features.
I try to use Nsight to see the performance, here is the result.
![one_batch_non_prefetch_time](https://user-images.githubusercontent.com/24785328/178502734-41532d17-9b5d-4a90-be8d-45bec641297b.png)
![one_batch_prefetch_cuda_time](https://user-images.githubusercontent.com/24785328/178502750-2dc97ea0-ab9f-48d3-b64c-59de3ce37402.png)
we can find that in each batch, when we use prefetch there need 326ms to get subgraph and correspanding features, and if we don't set prefetch, there only need 65+146=211ms to prepare batch data.

here is part of my code, any idea about this?:
```python
    # set to prefetch features
    # prefetch_src_data = {'paper': ['feat', 'label_one_hot'],
    #                      'author': ['feat'],
    #                      'field_of_study': ['feat'],
    #                      'institution': ['feat']}
    # prefetch_dst_data = {'paper':['label', 'label_emb']}
    # for i in range(args.NARS_R + 1):
    #     prefetch_dst_data['paper'].append('nars_feat_hop_' + str(i))

    
    prefetch_src_data = None
    prefetch_dst_data = None
    train_neighbor_sampler = dgl.dataloading.NeighborSampler(fanouts=train_fanouts,
                                                             prefetch_node_feats=prefetch_src_data,
                                                             prefetch_labels=prefetch_dst_data)

    torch.cuda.cudart().cudaProfilerStart()
    with torch.autograd.profiler.emit_nvtx():
        nvtx.range_push("train_loader")
        for _, seeds, blocks in train_dataloader:
            nvtx.range_pop()
            nvtx.range_push("input_data")

            features = blocks[0].srcdata['feat']
            feats_with_hops = []
            for i in range(nars_num_hops):
                feats_with_hops.append(blocks[-1].dstdata['nars_feat_hop_' + str(i)]['paper'])
            batch_label_emb = blocks[-1].dstdata['label_emb']['paper']

            nvtx.range_pop()
            nvtx.range_push("model")

            predictions = model(blocks, features, feats_with_hops, batch_label_emb,
                                node_type_flag, node_label_flag, edge_type_flag)

            cur_labels = blocks[-1].dstdata['label']['paper']
            loss = F.cross_entropy(predictions, cur_labels).to(device)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            train_correct += (predictions.argmax(1) == cur_labels).sum().item()
            train_total += predictions.shape[0]

            nvtx.range_pop()
            nvtx.range_push("train_loader")
    torch.cuda.cudart().cudaProfilerStop()
```

**Comment by yaox12** [comment]
From nsys profile, we can find that when you use prefetch, there are 3 big `IndexSelectMultiKernelAligned` kernels, which are DGL's optimized UVA indexing kernel. While without prefetch, there are only 2. Are you loading the same features in the two cases?

**Comment by maqy1995** [comment]
> @maqy1995 From the nsys profile, we can find that when you use prefetch, there are 3 big `IndexSelectMultiKernelAligned` kernels, which are DGL's optimized UVA indexing kernel. While without prefetch, there are only 2. Are you loading the same features in the two cases?

@yaox12 , you are right. When I use prefetch, there is an extra feat not used in the forward pass. After I fixed this, the performance became reasonable.
Thanks a lot!
